src/os/README.md

```diff
@@ -104,3 +108,125 @@ inspiration for several decisions in this library.
+constructor, a destructor and several setters, to manipulate an opaque C
+pointer. Instead of directly provide Hack version of these functions, a `shape`
+or `vec` corresponds to the opaque C pointer should be expose, and the underlying C
+struct should be internally created on demand from the `shape` or `vec`, when
+another underlying C function requires the opaque C pointer.
+
+### Unordered setters
+
+An opaque C pointers should map to a `shape` definition if the order to invoke
+setters does not matter. For example, given the following opaque C pointer and
+the related utility functions:
+
+``` c
+// The opaque C pointer
+typedef void *posix_spawnattr_t;
+
+// The constructor
+int posix_spawnattr_init(posix_spawnattr_t *attr);
+
+// The setters
+int posix_spawnattr_setpgroup(posix_spawnattr_t *attr, pid_t pgroup);
+
+#define	POSIX_SPAWN_RESETIDS 0x0001
+#define	POSIX_SPAWN_SETPGROUP 0x0002
+int posix_spawnattr_setflags(posix_spawnattr_t *attr, short flags);
+
+// The destructor
+int posix_spawnattr_destroy(posix_spawnattr_t *attr);
+```
+
+The corresponding Hack definition should be:
+
+``` hack
+newtype pid_t = int;
+enum PosixSpawnFlag : int {
+  POSIX_SPAWN_RESETIDS = 0x0001;
+  POSIX_SPAWN_SETPGROUP = 0x0002;
+}
+type posix_spawnattr_t = shape(
+  'posix_spawnattr_setpgroup' => pid_t,
+  'posix_spawnattr_setflags' => keyset<PosixSpawnFlag>,
+);
+```
+
+Note that the second parameter of `posix_spawnattr_setflags` is bit set,
+which maps to a `keyset` of `enum` in Hack.
+
+### Ordered repeatable setters
+
+An opaque C pointers should map to a `vec` definition if the setters can be
+invoked more than once and the order to invoke setters matters. For example,
+given the following opaque C pointer and the related utility functions:
+
+``` c
+// The opaque C pointer
+typedef void *posix_spawn_file_actions_t;
+
+// The constructor
+int posix_spawn_file_actions_init(posix_spawn_file_actions_t *file_actions);
+
+// The setters
+int posix_spawn_file_actions_addchdir(posix_spawn_file_actions_t *file_actions, const char *restrict path);
+int posix_spawn_file_actions_adddup2(posix_spawn_file_actions_t *file_actions, int filedes, int newfiledes);
+
+// The destructor
+int posix_spawn_file_actions_destroy(posix_spawn_file_actions_t *file_actions);
+```
+
+The corresponding Hack definition should be:
+
+``` hack
+newtype pid_t = int;
+enum PosixSpawnFlag : int {
+  POSIX_SPAWN_RESETIDS = 0x0001;
+  POSIX_SPAWN_SETPGROUP = 0x0002;
+}
+type posix_spawn_file_actions_t = vec[
+  shape(
+    'posix_spawn_file_actions_addchdir' => string,
+  )|
+  shape(
+    'posix_spawn_file_actions_adddup2' => shape(
+      'filedes' => FileDescriptor,
+      'newfiledes' => int,
+    ),
+  ),
+];
+```
+
+Since the original C function `posix_spawn_file_actions_adddup2` accepts three
+parameters, the second parameter and the third parameter are packed
+into a `shape`.
+
+Note that union types are not currently enabled for HSL. As a fallback encoding,
```

Comment:
There is not an expectation that they will be; we should not define rules that lead to a result that is not valid hack

Reply:
Updated